### DISCUSSION TITLE
Regexp freeze in user agent plugin for sdk-core consumes excessive CPU time?

### DISCUSSION THREAD

**Discussion by s-karki** [aws/aws-sdk-ruby#3004] [Q&A]
We run background job processors on a Fargate cluster, which pull jobs from an SQS queue (with Shoryuken). We've recently encountered an issue where the job processors' throughput starts to decrease when CPU usage reaches approximately 60%. Memory usage on the workers is normal, as is the time taken to fetch jobs off the queue. 

We ran a profiler to analyze this degradation in performance, and saw that [this code ](https://github.com/aws/aws-sdk-ruby/blob/e1786caffe124c49b78ed5b486225c453c26a34d/gems/aws-sdk-core/lib/aws-sdk-core/plugins/user_agent.rb#L133)to freeze a regex is taking up dramatically more CPU time during the degradation (occasionally up to 10x more). I was hoping to see if anyone has faced a similar issue, and what the potential root cause might be. 

I believe Regexes are immutable in Ruby – do we need freeze them? Or is there another fix for this? 

<img width="544" alt="image" src="https://github.com/aws/aws-sdk-ruby/assets/20472951/bbbb39da-0b2c-4bb5-ba3a-9061bf848001">

We started noticing this degradation in early March.

**Comment by mullermp** [comment]
Thanks for opening up an issue. This sounds strange. When there isn't a degradation (under 60% cpu usage), does this code functional normally and quickly?

Can you try a monkey patch to see if it would resolve your issue?

My thinking would be that, even when fixed, you're still reaching bottlenecks somehow, and some other block of code would become the problem. But the root cause seems like maybe your cluster is overloaded? Have you looked at scaling up your instances?
